ProcessTree.cpp

```diff
@@ -0,0 +1,530 @@
+    message.header.nlmsg_pid = getpid();
+    message.header.nlmsg_type = NLMSG_DONE;
+
+    message.connector.id.idx = CN_IDX_PROC;
+    message.connector.id.val = CN_VAL_PROC;
+    message.connector.len = sizeof(enum proc_cn_mcast_op);
+
+    message.mode = PROC_CN_MCAST_LISTEN;
+
+    if (send(_proc_socket, &message, sizeof(message), 0) < 0) {
+        Logger::Error("Cannot send to netlink socket for proc monitoring");
+        exit(1);
+    }
+}
+
+void ProcessNotify::run()
+{
+    struct __attribute__ ((aligned(NLMSG_ALIGNTO))) {
+        struct nlmsghdr header;
+        struct __attribute__ ((__packed__)) {
+            struct cn_msg connector;
+            struct proc_event event;
+        };
+    } message;
+
+    Logger::Info("ProcessNotify starting");
+
+    while(!IsStopping()) {
+        if (recv(_proc_socket, &message, sizeof(message), 0) <= 0) {
+            Logger::Error("Error receiving from netlink socket for process monitoring");
+        }
+
+        switch (message.event.what) {
+            case proc_event::what::PROC_EVENT_FORK:
+                _processTree->AddPnForkQueue(message.event.event_data.fork.child_pid, message.event.event_data.fork.parent_pid);
+                break;
+            case proc_event::what::PROC_EVENT_EXEC:
+                _processTree->AddPnExecQueue(message.event.event_data.exec.process_pid);
+                break;
+            case proc_event::what::PROC_EVENT_EXIT:
+                _processTree->AddPnExitQueue(message.event.event_data.exit.process_pid);
+                break;
+        }
+    }
+}
+
+void ProcessTree::AddPnForkQueue(int pid, int ppid)
+{
+    struct ProcessQueueItem p = {ProcessQueueFork, ProcessTreeSource_pnotify, pid, ppid};
+    std::unique_lock<std::mutex> queue_push_lock(_queue_push_mutex);
+    _PnQueue.push(p);
+    queue_push_lock.unlock();
+    std::unique_lock<std::mutex> queue_pop_lock(_queue_pop_mutex);
+    _queue_data_ready = true;
+    _queue_data.notify_one();
+}
+
+void ProcessTree::AddPnExecQueue(int pid)
+{
+    struct ProcessQueueItem p = {ProcessQueueExec, ProcessTreeSource_pnotify, pid};
+    std::unique_lock<std::mutex> queue_push_lock(_queue_push_mutex);
+    _PnQueue.push(p);
+    queue_push_lock.unlock();
+    std::unique_lock<std::mutex> queue_pop_lock(_queue_pop_mutex);
+    _queue_data_ready = true;
+    _queue_data.notify_one();
+}
+
+void ProcessTree::AddPnExitQueue(int pid)
+{
+    struct ProcessQueueItem p = {ProcessQueueExit, ProcessTreeSource_pnotify, pid};
+    std::unique_lock<std::mutex> queue_push_lock(_queue_push_mutex);
+    _PnQueue.push(p);
+    queue_push_lock.unlock();
+    std::unique_lock<std::mutex> queue_pop_lock(_queue_pop_mutex);
+    _queue_data_ready = true;
+    _queue_data.notify_one();
+}
+
+void ProcessTree::run()
+{
+    std::unique_lock<std::mutex> queue_pop_lock(_queue_pop_mutex);
+    while (!IsStopping()) {
+        _queue_data.wait(queue_pop_lock, [&]{return _queue_data_ready;});
```

Comment:
Would it not be simpler to use "!_PnQueue.empty()" instead of a separate bool?

Reply:
Yep, good idea; implemented.